server/schedule/placement/region_rule_cache.go

```diff
@@ -77,63 +71,162 @@ func (manager *RegionRuleFitCacheManager) CheckAndGetCache(region *core.RegionIn
+	return cache.isRegionUnchanged(region) && rulesEqual(cache.rules, rules) && storesEqual(cache.regionStores, stores)
 }
 
 func (cache *RegionRuleFitCache) isRegionUnchanged(region *core.RegionInfo) bool {
-	// we only cache region when it doesn't have down peers
-	if len(region.GetDownPeers()) > 0 || region.GetLeader() == nil {
+	// we only cache region when it is valid
+	if !validateRegion(region) {
 		return false
 	}
-	return region.GetLeader().StoreId == cache.region.GetLeader().StoreId &&
-		regionEpochEqual(cache.region, region)
+	return region.GetLeader().StoreId == cache.region.leaderStoreID &&
+		cache.region.epochEqual(region)
 }
 
-func regionEpochEqual(a, b *core.RegionInfo) bool {
-	if a.GetRegionEpoch() == nil || b.GetRegionEpoch() == nil {
+func rulesEqual(ruleCaches []ruleCache, rules []*Rule) bool {
+	if len(ruleCaches) != len(rules) {
 		return false
 	}
-	return a.GetRegionEpoch().Version == b.GetRegionEpoch().Version &&
-		a.GetRegionEpoch().ConfVer == b.GetRegionEpoch().ConfVer
+	return slice.AllOf(ruleCaches, func(i int) bool {
+		return slice.AnyOf(rules, func(j int) bool {
+			return ruleCaches[i].ruleEqual(rules[j])
+		})
+	})
 }
 
-func rulesEqual(a, b []*Rule) bool {
+func storesEqual(a []storeCache, b []*core.StoreInfo) bool {
 	if len(a) != len(b) {
 		return false
 	}
 	return slice.AllOf(a, func(i int) bool {
 		return slice.AnyOf(b, func(j int) bool {
-			return equalRules(a[i], b[j])
+			return a[i].storeEqual(b[j])
 		})
 	})
 }
 
-func storesEqual(a, b []*core.StoreInfo) bool {
-	if len(a) != len(b) {
+func validateRegion(region *core.RegionInfo) bool {
+	if region == nil || region.GetLeader() == nil || len(region.GetDownPeers()) > 0 || region.GetRegionEpoch() == nil {
 		return false
 	}
-	return slice.AllOf(a, func(i int) bool {
-		return slice.AnyOf(b, func(j int) bool {
-			return a[i].GetID() == b[j].GetID() &&
-				a[i].IsEqualLabels(b[j].GetLabels()) &&
-				a[i].GetState() == b[j].GetState()
+	return true
+}
+
+func toRegionRuleFitCache(region *core.RegionInfo, fit *RegionFit) *RegionRuleFitCache {
+	return &RegionRuleFitCache{
+		region:       toRegionCache(region),
+		regionStores: toStoreCacheList(fit.regionStores),
+		rules:        toRuleCacheList(fit.rules),
+		bestFit:      fit,
+	}
+}
+
+type ruleCache struct {
+	id       string
+	group    string
+	version  uint64
+	createTS uint64
+}
+
+func (r ruleCache) ruleEqual(rule *Rule) bool {
+	if rule == nil {
+		return false
+	}
+	return r.id == rule.ID && r.group == rule.GroupID && r.version == rule.Version && r.createTS == rule.CreateTimestamp
+}
+
+func toRuleCacheList(rules []*Rule) (c []ruleCache) {
+	for _, rule := range rules {
+		c = append(c, ruleCache{
+			id:       rule.ID,
+			group:    rule.GroupID,
+			version:  rule.Version,
+			createTS: rule.CreateTimestamp,
 		})
+	}
+	return c
+}
+
+type storeCache struct {
```

Comment:
prefer to use pointer receiver

Reply:
Why use pointer here? I tend to use struct in order to avoid panic error